sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1167,7 +1169,8 @@ class Analyzer(
           case p: Project =>
             val maybeResolvedExprs = exprs.map(resolveExpression(_, p))
             val (newExprs, newChild) = resolveExprsAndAddMissingAttrs(maybeResolvedExprs, p.child)
-            val missingAttrs = AttributeSet(newExprs) -- AttributeSet(maybeResolvedExprs)
+            // Only add missing attributes coming from `newChild`.
+            val missingAttrs = (AttributeSet(newExprs) -- p.outputSet).intersect(newChild.outputSet)
```

Comment:
This is a second time, but we need to fix in `Aggregate` case? The logic seems completely different. Or can we remove `Aggregate` case if `ResolveAggregateFunctions` can handle this? I don't think we have any reason to keep a wrong logic.

Reply:
Thanks. I think it's better to have a re-producible test case before changing `Aggregate` case. I'm trying to create a test case for it. Then it can be more confident to change `Aggregate` case.

Actually I found another place we need to fix. Seems we don't have enough test coverage for similar features.

Reply:
The logic gets convoluted here and we need to add comments. Basically we need to explain when we should expand the project list.

Reply:
The logic gets convoluted here and we need to add comments. Basically we need to explain when we should expand the project list.